streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -832,10 +837,103 @@
+     * "&lt;name&gt;" is an internally generated name, and "-repartition" is a fixed suffix.
+     * <p>
+     * The provided {@link KeyValueMapper} is applied to each input record and computes a new key for it.
+     * Thus, an input record {@code <K,V>} can be transformed into an output record {@code <K':V>}.
+     *
+     * @param keySelector      a {@link KeyValueMapper} that computes a new key for each record
+     * @return a {@code KStream} that contains records with new key (possibly of different type) and unmodified value
+     * @see #through(String)
+     * @see #through(String, Produced)
+     */
+    <KR> KStream<KR, V> repartition(final KeyValueMapper<? super K, ? super V, ? extends KR> keySelector);
+
+    /**
+     * Materialize this stream to an auto-generated repartition topic and create a new {@code KStream}
+     * from the auto-generated topic using {@link Serde key serde}, {@link Serde value serde}, {@link StreamPartitioner},
+     * number of partitions, and topic name part as defined by {@link Repartitioned}.
+     * <p>
+     * This operation is similar to {@link #through(String)}, however, Kafka Streams manages the used topic automatically.
+     * The Created topic is considered as an internal topic and is meant to be used only by current Kafka Streams instance.
+     * Similar to auto-repartitioning, the topic will be created with infinite retention time and data will be purged by Kafka Streams explicitly.
+     * The topic will be named as "${applicationId}-&lt;name&gt;-repartition", where "applicationId" is user-specified in
+     * {@link StreamsConfig} via parameter {@link StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG},
+     * "&lt;name&gt;" is either provided via {@link Repartitioned#as(String)} or an internally
+     * generated name, and "-repartition" is a fixed suffix.
+     * <p>
+     * The provided {@link KeyValueMapper} is applied to each input record and computes a new key for it.
+     * Thus, an input record {@code <K,V>} can be transformed into an output record {@code <K':V>}.
+     *
+     * @param keySelector      a {@link KeyValueMapper} that computes a new key for each record
+     * @param repartitioned the {@link Repartitioned} instance used to specify {@link Serdes},
+     *                      {@link StreamPartitioner} which determines how records are distributed among partitions of the topic,
+     *                      part of the topic name, and number of partitions for a repartition topic.
+     * @return a {@code KStream} that contains records with new key (possibly of different type) and unmodified value
+     * @see #through(String)
+     * @see #through(String, Produced)
+     */
+    <KR> KStream<KR, V> repartition(final KeyValueMapper<? super K, ? super V, ? extends KR> keySelector,
+                                    final Repartitioned<KR, V> repartitioned);
```

Comment:
Hi @vvcephei 
As far as I remember we didn't have any specific discussion around this operation. Main reason why we have added it I think is because of the convenience (merging selectKey and repartition into single operation). Similarly how `groupBy(final KeyValueMapper<? super K, ? super V, KR> selector, final Grouped<KR, V> grouped)` does it.

Reply:
I see. IIRC, your initial thought was to replace `groupBy` with `repartition`, but we've gotten away from that design. Now, it's more like a managed-topic version of `through` (in fact, this is what the javadoc for the method says). Maybe this is why I was confused to see this overload, since it makes less sense to think of changing the key at the last minute before `though` or `to`.

Are you particularly attached to this convenience overload? I'm just thinking it's a safer bet to add it later if people really want it than to add it now and never really know if it's useful or not.

Reply:
yup, agreed, fair point. removed.

Reply:
I can't say I am particulary attached to it, but I think it's useful one. From personal experience, we, at our company often do following topology:

```
streamBuilder.stream("input-topic")
.selectKey(...)
.through(...)
.transform(...) // transform is stateful 
```
here we need to explicitly use through in order to trigger repartitioning by selecting key (and we have to manage topic that we create in advance for `through` operation)

New way of doing same thing is quite nice I think:

```
streamBuilder.stream("input-topic")
.repartition(...) // select key here
.transform(...) // transform is stateful 
```

For us, it's quite common use-case. And.my _guess_ is it will be common use-case for anyone using complex `transform` operations in DSL. Does this make sense? On the other hand, if you believe that it's safer to remove `KeyValueMapper` overloads, I'll definitely do it.

Reply:
I certainly agree that it would be common to do a `selectKey` before a `repartition`. There's a trade-off to strike between the number of different operations you need and the number of options on a single operation you have to choose from.

I guess my hesitation is that it's still two different operations. For example, it also seems like it would be common to do a `map` before a `repartition`, but it's clearly now too much piled on if we have a fourth `repartition` overload also folding in the `map` operation. When I imagine coming to the API as a user, especially for the first time, I worry that I'd already have a lot of documentation to read to understand the implications of `repartition`, and each new overload adds linearly to the amount I have to learn to use the API. Plus, I just feel like I would be puzzled about the exact same question I asked above: is this overload just the same as `selectKey().repartition()`, or does it do something subtly different? As you can tell, I worry quite a bit about how we can make sure the API stays as simple as possible while we still add new functionality.

I guess this is just a long-winded way of saying that, yes, I would prefer to remove it :) Hopefully, this isn't too disappointing for you, since the main motivation was to save on managing the `through` topic, not necessarily to save on that extra `selectKey` operator.

Reply:
Thanks John, that's a valid point. Agree, I'll remove it.

Reply:
Done. I'll resolve this conversation.

Reply:
@lkokhreidze Can you update the KIP wiki page accordingly and send an follow up email to the VOTE thread of the KIP to highlight the change as an FYI that the KIP was modified (just in case somebody would have an objection, what I don't expect -- it's just custom in the community to do this).

Reply:
@mjsax yes, thanks for reminding me. Was meaning to do it.

Reply:
@mjsax done